### PR TITLE
feat(ExchangeList): Add Exchange._update() and ExchangeList.update_exchanges() methods

### PR DESCRIPTION
**Pull request by VIKTORVAV99**
## Issue

We want to be able to collect the most up to date exchange forecasts from ENTSO-E (an possibly others).

## Description
I went with a slightly different approach in this PR by adding dunder functions to the class. It results in slightly cleaner code (IMO) but could be harder to understand.

I'll let you guys decide if this is preferable over the method in https://github.com/electricitymaps/electricitymaps-contrib/pull/6442 otherwise I'll change it to use the same approach as there.

### Double check

- [x] I have run `pnpx prettier@2 --write .` and `poetry run format` in the top level directory to format my changes.

### CODE REVIEW

File: electricitymap/contrib/lib/models/event_lists.py
Comment (VIKTORVAV99) [inline electricitymap/contrib/lib/models/event_lists.py:152]: These can potentially be moved up to the parent classes and be re-used between all the parser class lists if we want to.

File: electricitymap/contrib/lib/models/event_lists.py
Comment (VIKTORVAV99) [inline electricitymap/contrib/lib/models/event_lists.py:223]: With the changes in the follow up PR it will look like this instead:
```python
        for new_event in new_exchanges:
            if new_event in exchanges:
                existing_event = exchanges[new_event]
                updated_event = Exchange.update(existing_event, new_event)
                exchanges[new_event] = updated_event
            else:
                exchanges._append_event(new_event)
```
Which could be further reduced by removing the intermediate variables if we want.

**Review by pierresegonne — COMMENTED** [review]
I think it can be a good idea to use the magic methods, but at the condition that we ensure that we are consistent with other Implementations. 

Should we do it in this PR? Maybe not, it's a bit frustrating, but if we truly want to separate concerns, it would be better to just add the functionality in this PR and follow up with another one to transition to magic methods throughout. 

In this case, I'd be ok going forward with magic methods in this PR, but we need to be 100% sure that we'll follow-up afterwards to ensure consistency

File: electricitymap/contrib/lib/models/event_lists.py
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:223]: append is not a magic method in it self and to avoid too much logic in the magic methods this made the most sense imo, but I have made follow up changes locally.

**Comment by VIKTORVAV99** [comment]
> I think it can be a good idea to use the magic methods, but at the condition that we ensure that we are consistent with other Implementations.
> 
> Should we do it in this PR? Maybe not, it's a bit frustrating, but if we truly want to separate concerns, it would be better to just add the functionality in this PR and follow up with another one to transition to magic methods throughout.
> 
> In this case, I'd be ok going forward with magic methods in this PR, but we need to be 100% sure that we'll follow-up afterwards to ensure consistency

I'll push my local branch as a draft PR so you can evaluate how I have done in there, that moves the magic methods to the base classes and adds appropriate overrides in all the child functions for proper typing.

**Comment by unitrium** [comment]
I started reviewing this but I won't have enough time to finish it today, if you don't mind waiting for me, I will give you feedback tomorrow.

File: electricitymap/contrib/lib/models/event_lists.py
Comment (unitrium) [reply electricitymap/contrib/lib/models/event_lists.py:152]: Those changes make sense to me, but I think they should be in the parent class rather.

File: electricitymap/contrib/lib/models/events.py
Comment (unitrium) [inline electricitymap/contrib/lib/models/events.py:434]: Not sure I fully understand why this would be usef by someone if they already created an Exchange.
Maybe this should be an internal method of the list rather?
I'm not sure someone would understand what this is for.

File: electricitymap/contrib/lib/models/event_lists.py
Comment (unitrium) [inline electricitymap/contrib/lib/models/event_lists.py:212]: Hum, I'm thinking that the datetime indexing is not super explicit, since this is initially a list of events.
This should probably be explicited in the class description.

**Review by unitrium — CHANGES_REQUESTED** [review]
Overall I think the change is good, I think we should just abstract it more and do it in the parent class.
Otherwise we have a very different behaviour for Exchanges and other types.

File: electricitymap/contrib/lib/models/event_lists.py
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:212]: That is what I do in the follow up PR, but I can do it directly in here as well and then update the other PR to just contain the logic changes I want to make.

File: electricitymap/contrib/lib/models/event_lists.py
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:152]: That is what I do in the follow up PR, but I can do it directly in here as well and then update the other PR to just contain the logic changes I want to make.

File: electricitymap/contrib/lib/models/event_lists.py
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:212]: In my follow up PR I am changing this to be a dict internally to streamline a lot of the logic without changing how it works externally.

It drastically simplifies the logic of all the magic functions.

That PR is still a draft though as it hasn't beef finalized or fully tested yet.

File: electricitymap/contrib/lib/models/events.py
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/events.py:434]: The reason this is in here is because it vastly simplifies the logic in the lists and can be abstracted away (could even be possible to do in the top level event_list if all the events types have a update function).

But I'll prepend it weith `_`to mark it as private.

File: electricitymap/contrib/lib/models/event_lists.py
Comment (VIKTORVAV99) [reply electricitymap/contrib/lib/models/event_lists.py:220]: Added a it to the top level EventList class as there is where the magic functions are implemented now. Let me know if you want to make it even more clear.

**Comment by unitrium** [comment]
Will review a bit later during the week.

**Comment by pierresegonne** [comment]
I'll let @unitrium do the final review

**Review by unitrium — APPROVED** [review]
Looks good to me! Thanks for adding this!

**Pull request merged**
